### PR TITLE
Workaround for buggy PKCS#11 C_Sign() implementation

### PR DESCRIPTION
**Pull request by mskalski**
Add workaround for buggy implementation of PKCS#11 library when it
does not update signature_length parameter whein it returns CKR_BUFFER_TOO_SMALL.
It handles also (hypotetical) reducing by PKCS#11 library returned signature length.

Additionally speeds up signature using larger buffer (of 1024 bytes) at beginning
which should result for most of cases calling C_Sign() only once.

### CODE REVIEW

**Comment by wolneykien** [comment]
Hi! Thanks, I'll look at it tonight.

**Comment by wolneykien** [comment]
I think we have to limit the possible number of iterations.

**Comment by wolneykien** [comment]
And it would be nice to know the signature length beforehand. I think it is possible. At least for OpenSSL...

**Comment by mskalski** [comment]
> And it would be nice to know the signature length beforehand. I think it is possible. At least for OpenSSL...

Yes it is possible for RSA (size of RSA key), of course it is possible for Elliptic curve, (but you need to know key type and curve), but unnecessary.

I think buffer of size 1024 bytes is far enough for all types of signatures and usually causes only one call to C_Sign().

File: src/common/pkcs11_lib.c [hunk 1]
```diff
@@ -1811,8 +1811,9 @@ int sign_value(pkcs11_handle_t *h, cert_object_t *cert, CK_BYTE *data,
     return -1;
   }
   *signature = NULL;
-  *signature_length = 64;
+  *signature_length = 1024;
```
Comment (wolneykien) [inline src/common/pkcs11_lib.c:1814]: Why do we need a larger initial buffer length here? There is the increment step below:

`*signature_length = current_signature_length * 2;`

**Review by wolneykien — CHANGES_REQUESTED** [review]
Please, limit the number of possible iterations — or we would still have the endless loop there. May be we should define a maximum possible buffer size?

File: src/common/pkcs11_lib.c [hunk 2]
```diff
@@ -1824,6 +1825,11 @@ int sign_value(pkcs11_handle_t *h, cert_object_t *cert, CK_BYTE *data,
       /* increase signature length as long as it it to short */
       free(*signature);
       *signature = NULL;
+      if (current_signature_length >= *signature_length) {
+        /* workaround for buggy PKCS#11 implementation: it didn't change
+           or even lowered buffer size - forcing using larger (double size) buffer */
+        *signature_length = current_signature_length * 2;
+      }
```
Comment (mskalski) [inline src/common/pkcs11_lib.c:1832]: Well, theoretically we could have here endless loop, but only in the case when this buggy PKCS#11 library returns always `CKR_BUFFER_TOO_SMALL` even for large buffer sizes. I mean PKCS#11 library did check size and returned this error, so I think we can assume it was done intentionally. Bug here is not updating `signature_length` parameter, not intentional messing with returned data.

Yes, we can check maximum value here, but what size we can select? 65536 bytes?

**Comment by martinpaljak** [comment]
If it is a buggy module, maybe you can
- report it to the module vendor to be fixed
- mention the name/type/version of it so that could be tracked.

File: src/common/pkcs11_lib.c (hunk 2 above)
Comment (wolneykien) [reply src/common/pkcs11_lib.c:1832]: Yes, see https://github.com/OpenSC/pam_pkcs11/commit/7ce58a6c6cdd8f6b90fffe552d2b7137677da21a .

**Pull request merged**